index.js

```diff
@@ -1270,78 +947,15 @@
                    rejectInconsistent);
     };
 
-    //  paramTypesAndValues :: ((a -> Any), Type, [a]) -> ErrorRepr
-    var paramTypesAndValues = function(valFn, type, values) {
-      return map(
-        chain(values, valFn),
-        function(value) {
-          return typesAndValues(
-            type,
-            [value]
-          );
-        }
-      );
-    };
-
-    //  nullaryTypesAndValues :: [Any] -> ErrorRepr
-    var nullaryTypesAndValues = function(values) {
-      return {
-        isParameterizedType: false,
-        types: valueTypes(values),
-        values: values
-      };
-    };
-
-    //  typesAndValues :: (type, [Any]) -> ErrorRepr
-    var typesAndValues = function(t, values) {
-      if (isParameterizedType(t)) {
-        var constructorName = functionName(values[0].constructor);
```

Comment:
> Identity(42).constructor.name;
> // => 'Object'

Could check for @@type if constructor.name is Object?

> S.Just(42).constructor.name;
> // => 'Maybe'

The way that Maybe -> Just/Nothing and Either -> Left/Right are defined is pretty unique I would say. In the comments to plaid/sanctuary#123 you [wrote](https://github.com/plaid/sanctuary/pull/123#issuecomment-168406709):

> When defining a method for both Left and Right, say, it's convenient to share the constraints and the list of types, so it seems more appropriate to have a single definition with one internal branch than two separate definitions with no branches.

Another alternative is to partially apply `def` (and `method`), leaving off only the function definition. So take `Either` and `chain` for example:

```js
var Either_chain = method(
  'Either#chain',
  {},
  [$Either(a, b), $.Function, $Either(a, c)]
);

Right.prototype.chain = Either_chain(function(right, f) {
  return f(right.value);
});

Left.prototype.chain = Either_chain(function(left, f) {
  return left;
});
```
Maybe the best of both worlds? We don't need to `def` functions in multiple places and the actual function bodies become super simple again, without any type branching.

We could then go back to the more standard constructors before #123, the instances will actually be instances of the things they are instantiating, our error messages will tell us what we actually have, etc. And of course we don't need to bring back inheritance, although we could if we wanted?